projects/core/src/main/resources/data/computercraft/lua/rom/modules/main/cc/audio/wav.lua

```diff
@@ -0,0 +1,224 @@
+                retval.codec = "f32"
+            elseif format == 0xFFFE then
+                bitDepth = str_unpack("<H", chunk, 19)
+                local uuid = str_sub(chunk, 25, 40)
+                if uuid == wavExtensible.pcm then
+                    dataType = bitDepth == 8 and "unsigned" or "signed"
+                    retval.codec = (bitDepth == 8 and "u" or "s") .. bitDepth
+                elseif uuid == wavExtensible.dfpwm then
+                    dataType = "dfpwm"
+                    retval.codec = "dfpwm"
+                elseif uuid == wavExtensible.pcm_float then
+                    dataType = "float"
+                    retval.codec = "f32"
+                else error("unsupported WAV file", 2) end
+            else error("unsupported WAV file", 2) end
+        elseif temp == "data" then
+            local data = str_sub(data, pos, pos + size - 1)
+            if #data < size then error("invalid WAV file", 2) end
+            if not retval.length then retval.length = size / blockAlign end
+            pos = pos + size
+            local pos = 1
+            local channels = retval.channels
+            if dataType == "dfpwm" then
+                local decoder = dfpwm.make_decoder()
+                function retval.read(samples)
+                    if pos > #data then return nil end
+                    local chunk = decoder(str_sub(data, pos, pos + math.ceil(samples * channels / 8) - 1))
+                    pos = pos + math.ceil(samples * channels / 8)
+                    local res = {}
+                    for i = 1, channels do
+                        local c = {}
+                        res[i] = c
+                        for j = 1, samples do
+                            c[j] = chunk[(j - 1) * channels + i]
+                        end
+                    end
+                    return table.unpack(res)
+                end
+            else
+                local format = dataType == "unsigned" and "I" .. (bitDepth / 8) or (dataType == "signed" and "i" .. (bitDepth / 8) or "f")
+                local transform
+                if dataType == "unsigned" then
+                    function transform(n) return n - 128 end
+                elseif dataType == "signed" then
+                    if bitDepth == 16 then function transform(n) return math_floor(n / 0x100) end
+                    elseif bitDepth == 24 then function transform(n) return math_floor(n / 0x10000) end
+                    elseif bitDepth == 32 then function transform(n) return math_floor(n / 0x1000000) end end
+                elseif dataType == "float" then
+                    function transform(n) return math_floor(n * (n < 0 and 128 or 127)) end
+                end
+                function retval.read(samples)
+                    if pos > #data then return nil end
+                    local chunk = { ("<" .. format:rep(math.min(samples * channels, (#data - pos + 1) / (bitDepth / 8)))):unpack(data, pos) }
```

Comment:
Oh, I don't feel good about this line (the extra format string allocation, and then putting it into a table), but some quick benchmarks do show it's the quickest approach :(.

Reply:
Yeah, I didn't like it either, but I found the same results after a comparison in AUKit. It's gross, but it's for the best.